File: packages/ios/Sources/CivUI/CivCheckboxGroup.swift
// CivUI — CivCheckboxGroup for SwiftUI
// Accessible checkbox group following government design system patterns.
// Renders: legend → hint → error → [checkbox options] (Section 508 compliant)

import SwiftUI

/// Checkbox option definition for CivCheckboxGroup.
public struct CivCheckboxOption: Identifiable {
    public let id: String
    public let label: String
    public let value: String
    public var description: String?
    public var isDisabled: Bool

    public init(label: String, value: String, description: String? = nil, isDisabled: Bool = false) {
        self.id = value
        self.label = label
        self.value = value
        self.description = description
        self.isDisabled = isDisabled
    }
}

/// Accessible checkbox group for government applications.
///
/// Groups multiple checkboxes with a shared legend, hint, and error message.
/// Manages multi-select state via a `[String]` binding.
///
/// Usage:
/// ```swift
/// CivCheckboxGroup(
///     legend: "Select all that apply",
///     values: $selectedValues,
///     options: [
///         CivCheckboxOption(label: "Email updates", value: "email"),
///         CivCheckboxOption(label: "Text alerts", value: "sms"),
///         CivCheckboxOption(label: "Phone calls", value: "phone"),
///     ],
///     isRequired: true
/// )
/// ```
public struct CivCheckboxGroup: View {
    // MARK: - Properties

    /// Legend text for the group.
    public let legend: String

    /// Bound array of checked values.
    @Binding public var values: [String]

    /// Available checkbox options.
    public let options: [CivCheckboxOption]

    /// Help text shown below the legend.
    public var hint: String?

    /// Error message for the group.
    public var error: String?

    /// Whether at least one selection is required.
    public var isRequired: Bool

    /// Whether the group is disabled.
    public var isDisabled: Bool

    /// Whether the group is read-only (visible but non-interactive, no opacity change).
    public var isReadonly: Bool

    /// Field name for form state registration.
    public var formName: String

    /// Whether to render in tile (card) style.
    public var isTile: Bool

    /// Layout orientation for options: "vertical" (default) or "horizontal".
    public var orientation: String

    /// Whether to show a "Select all" option.
    public var showSelectAll: Bool

    /// Preset options key (e.g., "us-states").
    public var preset: String

    /// Preset variant for alternative option sets.
    public var presetVariant: String

    /// Maximum number of selections allowed. Nil = unlimited.
    public var maxSelections: Int?

    /// Minimum number of selections required. Nil = no minimum.
    public var minSelections: Int?

    /// Tile rendering variant. `auto` picks `card` for ≤4 options and
    /// `list` for 5+. Vertical + tile only.
    public var layout: String

    /// Called when the set of checked values changes.
    public var onChange: (([String]) -> Void)?

    /// Called for analytics tracking (parallels `civ-analytics` event).
    public var onAnalytics: ((String, [String: Any]?) -> Void)?

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        legend: String,
        values: Binding<[String]>,
        options: [CivCheckboxOption],
        hint: String? = nil,
        error: String? = nil,
        isRequired: Bool = false,
        isDisabled: Bool = false,
        isReadonly: Bool = false,
        formName: String = "",
        isTile: Bool = false,
        orientation: String = "vertical",
        showSelectAll: Bool = false,
        preset: String = "",
        presetVariant: String = "",
        maxSelections: Int? = nil,
        minSelections: Int? = nil,
        layout: String = "auto",
        onChange: (([String]) -> Void)? = nil,
        onAnalytics: ((String, [String: Any]?) -> Void)? = nil
    ) {
        self.legend = legend
        self._values = values
        self.options = options
        self.hint = hint
        self.error = error
        self.isRequired = isRequired
        self.isDisabled = isDisabled
        self.isReadonly = isReadonly
        self.formName = formName
        self.isTile = isTile
        self.orientation = orientation
        self.showSelectAll = showSelectAll
        self.preset = preset
        self.presetVariant = presetVariant
        self.maxSelections = maxSelections
        self.minSelections = minSelections
        self.layout = layout
        self.onChange = onChange
        self.onAnalytics = onAnalytics
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: CivTokens.Spacing._1) {
            // 1. Legend
            legendView

            // 2. Hint
            if let hint, !hint.isEmpty {
                Text(hint)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Base.dark,
                        dark: CivTokens.DarkColors.Base.dark
                    ))
                    .accessibilityIdentifier("civ-hint")
            }

            // 3. Error
            if let error, !error.isEmpty {
                Text(error)
                    .font(.system(size: CivTokens.Typography.FontSize.sm,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityIdentifier("civ-error")
            }

            // 4. Checkbox options
            optionsLayout
        }
        .padding(.bottom, CivTokens.Spacing._4)
        .accessibilityElement(children: .contain)
        .onChange(of: error) { newError in
            if let newError, !newError.isEmpty {
                UIAccessibility.post(notification: .announcement, argument: newError)
            }
        }
    }

    // MARK: - Options Layout

    @ViewBuilder
    private var optionsLayout: some View {
        if orientation == "horizontal" {
            HStack(alignment: .top, spacing: CivTokens.Spacing._2) {
                optionsContent
            }
        } else {
            VStack(alignment: .leading, spacing: CivTokens.Spacing._2) {
                optionsContent
            }
        }
    }

    @ViewBuilder
    private var optionsContent: some View {
        ForEach(options) { option in
            CivCheckbox(
                label: option.label,
                checked: checkedBinding(for: option.value),
                value: option.value,
                description: option.description,
                isDisabled: isDisabled || option.isDisabled,
                isReadonly: isReadonly,
                isTile: isTile,
                onChange: { checked, val in
                    if checked {
                        if !values.contains(val) {
                            values.append(val)
                        }
                    } else {
                        values.removeAll { $0 == val }
                    }
                    onChange?(values)
                    onAnalytics?("change", ["values": values])
                }
            )
        }
    }

    // MARK: - Helpers

    private func checkedBinding(for value: String) -> Binding<Bool> {
        Binding(
            get: { values.contains(value) },
            set: { newValue in
                if newValue {
                    if !values.contains(value) {
                        values.append(value)
                    }
                } else {
                    values.removeAll { $0 == value }
                }
            }
        )
    }

    // MARK: - Subviews

    private var legendView: some View {
        HStack(spacing: CivTokens.Spacing._0_5) {
            Text(legend)
                .font(.system(size: CivTokens.Typography.FontSize.base,
                              weight: CivTokens.Typography.FontWeight.bold))
                .foregroundColor(adaptiveColor(
                    light: CivTokens.Colors.Base.darkest,
                    dark: CivTokens.DarkColors.Base.darkest
                ))

            if isRequired {
                Text("*")
                    .font(.system(size: CivTokens.Typography.FontSize.base,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityLabel("required")
            }
        }
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivCheckboxGroup_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        @State private var notifications: [String] = ["email"]
        @State private var services: [String] = []

        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    CivCheckboxGroup(
                        legend: "Notification preferences",
                        values: $notifications,
                        options: [
                            CivCheckboxOption(label: "Email updates", value: "email"),
                            CivCheckboxOption(label: "Text alerts", value: "sms"),
                            CivCheckboxOption(label: "Phone calls", value: "phone"),
                        ],
                        hint: "Select all that apply",
                        isRequired: true
                    )

                    CivCheckboxGroup(
                        legend: "Additional services",
                        values: $services,
                        options: [
                            CivCheckboxOption(label: "Expedited processing", value: "expedited", description: "Additional fee applies"),
                            CivCheckboxOption(label: "Certified copy", value: "certified", description: "Notarized document"),
                        ],
                        isTile: true
                    )
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif

File: packages/android/src/main/kotlin/gov/civui/components/CivCheckboxGroup.kt
// CivUI — CivCheckboxGroup for Jetpack Compose
// Accessible checkbox group with legend, hint, and error support.
// Renders: legend -> hint -> error -> checkbox options (Section 508 compliant)

package gov.civui.components

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.i18n.CivLocale
import gov.civui.tokens.CivTokens

/**
 * Data class representing a checkbox option within a group.
 */
data class CivCheckboxOption(
    val value: String,
    val label: String,
    val description: String = "",
    val disabled: Boolean = false,
)

/**
 * Accessible checkbox group for government applications.
 *
 * Groups multiple checkboxes with a shared legend, hint, and error message.
 * Manages a list of selected values and reports changes via [onValuesChange].
 *
 * Usage:
 * ```kotlin
 * var selected by remember { mutableStateOf(listOf<String>()) }
 * CivCheckboxGroup(
 *     legend = "Select all that apply",
 *     values = selected,
 *     onValuesChange = { selected = it },
 *     options = listOf(
 *         CivCheckboxOption(value = "a", label = "Option A"),
 *         CivCheckboxOption(value = "b", label = "Option B"),
 *     ),
 * )
 * ```
 */
@Composable
fun CivCheckboxGroup(
    legend: String,
    values: List<String>,
    onValuesChange: (List<String>) -> Unit,
    options: List<CivCheckboxOption>,
    modifier: Modifier = Modifier,
    hint: String? = null,
    error: String? = null,
    tile: Boolean = false,
    orientation: CivOrientation = CivOrientation.Vertical,
    required: Boolean = false,
    disabled: Boolean = false,
    readonly: Boolean = false,
    name: String = "",
    showSelectAll: Boolean = false,
    preset: String = "",
    presetVariant: String = "",
    maxSelections: Int = 0,
    minSelections: Int = 0,
    onAnalytics: ((event: String, data: Map<String, Any>?) -> Unit)? = null,
) {
    val isDark = isSystemInDarkTheme()

    val labelColor = if (isDark) CivTokens.DarkColors.Base.darkest else CivTokens.Colors.Base.darkest
    val hintColor = if (isDark) CivTokens.DarkColors.Base.dark else CivTokens.Colors.Base.dark
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_

    Column(
        modifier = modifier
            .padding(bottom = CivTokens.Spacing._4)
            .then(
                if (error != null) Modifier.semantics { error(error) } else Modifier
            ),
    ) {
        // 1. Legend
        CivLabel(
            label = legend,
            required = required,
            labelColor = labelColor,
            errorColor = errorColor,
        )

        // 2. Hint
        CivHint(text = hint, color = hintColor)

        // 3. Error
        CivError(text = error, color = errorColor)

        // 4. Checkbox options
        val content: @Composable () -> Unit = {
            options.forEach { option ->
                CivCheckbox(
                    label = option.label,
                    checked = values.contains(option.value),
                    onCheckedChange = { checked ->
                        val newValues = if (checked) {
                            values + option.value
                        } else {
                            values - option.value
                        }
                        onValuesChange(newValues)
                        onAnalytics?.invoke("change", mapOf("field" to legend, "values" to newValues))
                    },
                    value = option.value,
                    description = option.description,
                    tile = tile,
                    disabled = disabled || option.disabled,
                    readonly = readonly,
                )
            }
        }

        if (orientation == CivOrientation.Horizontal) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(CivTokens.Spacing._4),
            ) {
                content()
            }
        } else {
            Column {
                content()
            }
        }
    }
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivCheckboxGroup")
@Composable
private fun CivCheckboxGroupPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        var selected by remember { mutableStateOf(listOf("email")) }
        CivCheckboxGroup(
            legend = "How should we contact you?",
            values = selected,
            onValuesChange = { selected = it },
            options = listOf(
                CivCheckboxOption(value = "email", label = "Email"),
                CivCheckboxOption(value = "phone", label = "Phone"),
                CivCheckboxOption(value = "mail", label = "Mail", description = "Physical mail to your address on file"),
            ),
            required = true,
            hint = "Select all that apply",
        )

        var tileSelected by remember { mutableStateOf(listOf<String>()) }
        CivCheckboxGroup(
            legend = "Document types needed",
            values = tileSelected,
            onValuesChange = { tileSelected = it },
            options = listOf(
                CivCheckboxOption(value = "passport", label = "Passport"),
                CivCheckboxOption(value = "license", label = "Driver's license"),
            ),
            tile = true,
            error = "Select at least one document type",
        )
    }
}
